File: leetcode/0001_two-sum/solution.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	res := []int{-1, -1}
	numsMap := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		if v, ok := numsMap[target-nums[i]]; ok {
			res[0] = v
			res[1] = i
			break
		}
		numsMap[nums[i]] = i
	}
	return res
}

func main() {
	res := twoSum([]int{2, 7, 11, 15}, 9)
	fmt.Println(res)
}

File: leetcode/0001_two-sum/solution.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* twoSum(int* nums, int numsSize, int target)
{
    int* res = malloc(sizeof(int) * 2);
    for (int i = 0; i < numsSize - 1; ++i) {
        int x = target - nums[i];
        for (int j = i+1; j < numsSize; ++j) {
            if (x == nums[j]) {
                res[0] = i;
                res[1] = j;
                return res;
            }
        }
    }
    res[0] = -1;
    res[1] = -1;
    return res;
}

/* 测试 */
int main()
{
    int nums[] = {2, 7, 11, 15};
    int* res = twoSum(nums, 4, 9);
    printf("[%d, %d]\n", res[0], res[1]);
    free(res);
    return 0;
}
